upup/pkg/kutil/rollingupdate_cluster.go

```diff
@@ -269,35 +281,169 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
+
+	if isBastion || rollingUpdateData.CloudOnly {
+		glog.Infof("Not validating the cluster")
+	} else if rollingUpdateData.FailOnValidate && featureflag.DrainAndValidateRollingUpdate.Enabled() {
+		if err = n.ValidateCluster(rollingUpdateData, instanceGroupList); err != nil {
+			glog.Errorf("Error validating cluster: %s.", err)
+			return err
+		}
+	}
+
 	for _, u := range update {
-		instanceID := aws.StringValue(u.ASGInstance.InstanceId)
-		glog.Infof("Stopping instance %q in AWS ASG %q", instanceID, n.ASGName)
 
-		// TODO: Evacuate through k8s first?
+		if isBastion {
+
+			if err = n.DeleteAWSInstance(u, c); err != nil {
+				return err
+			}
+			glog.Infof("Updated a bastion instance and continuing.")
+
+			continue
+		}
+
+		if rollingUpdateData.CloudOnly {
+			glog.Warningf("Not draining nodes - cloud only is set.")
+		} else if featureflag.DrainAndValidateRollingUpdate.Enabled() {
+
+			glog.Infof("Draining the node.")
+
+			if err = n.DrainNode(u, rollingUpdateData); err == nil {
+				glog.Errorf("Error draining node: %s.", err)
+				return err
+			}
+		}
+
+		if err = n.DeleteAWSInstance(u, c); err != nil {
+			glog.Errorf("Error deleting aws instance: %s.", err)
+			return err
+		}
+
+		// Wait for new EC2 instances to be created
+		time.Sleep(t)
+
+		if rollingUpdateData.CloudOnly {
+			glog.Warningf("Not validating nodes as cloudonly flag is set.")
+			return nil
+		} else if featureflag.DrainAndValidateRollingUpdate.Enabled() {
+
+			glog.Infof("Validating the cluster.")
+			if err = n.ValidateClusterWithRetries(rollingUpdateData, instanceGroupList, t); err != nil {
+				glog.Errorf("Error validating cluster: %s.", err)
+				return err
+			}
+
+			if err != nil && rollingUpdateData.FailOnValidate {
+				return fmt.Errorf("validation timed out while performing rolling update: %v", err)
+			}
+		}
+
+	}
+
+	return nil
+}
+
+// ValidateClusterWithRetries runs our validation methods on the K8s Cluster x times and then fails.
+func (n *CloudInstanceGroup) ValidateClusterWithRetries(rollingUpdateData *RollingUpdateCluster, instanceGroupList *api.InstanceGroupList, t time.Duration) (err error) {
+
+	for i := 0; i <= rollingUpdateData.ValidateRetries; i++ {
+		_, err = validation.ValidateCluster(rollingUpdateData.ClusterName, instanceGroupList, rollingUpdateData.K8sClient)
+		if err != nil {
+			glog.Infof("Waiting longer for kops validate to pass: %s.", err)
+			time.Sleep(t / 2)
+		} else {
+			glog.Infof("Cluster validated proceeding with next step in rolling update.")
+			break
+		}
+	}
+
+	return nil
+}
+
+// ValidateCluster runs our validation methods on the K8s Cluster.
+func (n *CloudInstanceGroup) ValidateCluster(rollingUpdateData *RollingUpdateCluster, instanceGroupList *api.InstanceGroupList) error {
+
+	_, err := validation.ValidateCluster(rollingUpdateData.ClusterName, instanceGroupList, rollingUpdateData.K8sClient)
+	if err != nil {
+		return fmt.Errorf("cluster %s does not pass validation", rollingUpdateData.ClusterName)
```

Comment:
Include the inner error?  Also I don't think anyone is going to need the cluster name?